### PR TITLE
LOG-1441: Easy, accessible build and install.

### PR DESCRIPTION
**Pull request by alanconway**
NOTE: Make targets have changed, see `make help` and docs/HACKING.md.

IMPORTANT:
- REQUIRES A PUBLIC IMAGE REGISTRY for local build & test, e.g. on quay.io
- Must do `make overlay REGISTRY=quay.io/myname` to initialize a developer overlay.
- Definition of images, namespace etc. in kustomize overlays, not env variables.
  - but still respects CI env variables when run under CI.
- Use operator-sdk and kustomize instead of custom scripts.

Less important:

- Quick deployment without OLM `make deploy`, with OLM `make olm-deploy`
- Make automatically (re)builds required out-dated dependencies,
  - e.g. `make deploy` builds & deploys everything  from a fresh checkout.

FYI: Reasons for not using the cluster-internal registry:
- More realistic, this is how images are normally delivered.
- Port-forwarding code for internal registry is racy - bogus error messages, flakes.
- Internal registry is confusing because:
  - Different names for the same image inside/outside cluster.
  - Pushing to the internal registry seems significant, but isn't;
    we're just using it as a registry "because it's there"

```
make help
MAKE TARGETS
  help             Display this help
  all              Build everything, run all tests. Requires a cluster.
Build the operator
  init             Initialize your workspace for local build & test, must set REGISTRY.
  clean            Delete build artifacts and temporary files, keep downloaded tools and overlay.
  spotless         Delete everything including tools and overlay.
  image            Build the operator image
  image-push       Push the operator image
Deploy directly (without OLM)
  deploy           Deploy kustomized resources.
  undeploy         Delete deployed resources from cluster.
Deploy via OLM (Operater Lifecycle Manager)
  bundle-image     Build the OLM bundle image.
  bundle-push      Push the OLM bundle image.
  olm-deploy       Deploy the OLM bundle.
Testing
  check            Checks that don't need a cluster: generate, compile, lint, unit test.
  env              Environment for running tests.
  run              Run the operator outside the cluster for debugging.
  test-unit        Run unit tests that don't need a cluster.
  test-functional  Run tests that require a cluster, but not a deployment.
  deploy-elastic   Deploy elasticsearch for e2e tests.
  undeploy-elastic  Undeploy elasticsearch.
  test-e2e         Run e2e tests, asssumes operator and elasticsearch already deployed
  test-e2e-olm     Deploy and run e2e tests from scratch for CI.
  test-clean       Delete left-over test namespaces in the cluster
QUICK START
  make init REGISTRY=quay.io/YOUR_QUAY_ACCOUNT_NAME
  make deploy
RUNNING INDIVIDUAL TESTS
  export $(make env)
  LOG_LEVEL=2 go test -v test/functional/outputs
  make deploy deploy-elastic
  go test ./test/e2e/logforwarding/loki/forward_to_loki_test.go
```

### CODE REVIEW

**Comment by vimalk78** [comment]
please add a WIP label if applicable

File: Dockerfile.in
Comment (periklis) [inline Dockerfile.in:58]: We need to double check this on midstream/downstream. IIRC until 5.1 we used the CLO container image built by brew to extract the manifests and compile the bundle container with brew. I believe we don't do this anymore and we don't need these lines anymore.

File: config/manager/kustomization.yaml
Comment (periklis) [inline config/manager/kustomization.yaml:7]: This is one of the changes we need to watch out and never commit.

**Review by periklis — CHANGES_REQUESTED** [review]
This one goes into the right direction to modernize this repository and make use of `operator-sdk` as the main and single driver for daily development tasks. However ignoring generated files from Git and specifically the bundle directory and `bundle.Dockerfile` will break two important pieces in our CI/CD workflow:
- Removing them will break the upstream CI building the bundler container (See [here](https://github.com/openshift/release/blob/master/ci-operator/config/openshift/cluster-logging-operator/openshift-cluster-logging-operator-master.yaml#L83) and [docs](https://docs.ci.openshift.org/docs/how-tos/testing-operator-sdk-operators/)). We use this extract job to lint that our bundle container images never break on upstream.
- Removing them will break the midstream process to process and generate the bundle contents for dist-git (See [midstream render_templates](https://code.engineering.redhat.com/gerrit/plugins/gitiles/openshift-logging/+/refs/heads/openshift-logging-5.3-rhel-8/distgit/containers/cluster-logging-operator-bundle/render_templates)).

**Comment by alanconway** [comment]
/hold
wait for @vparfonov to merge #1240 which overlaps with this. 
Need to rebase the Makefile and registry improvements on his operator-sdk and RBAC changes.

File: Dockerfile.in
Comment (alanconway) [reply Dockerfile.in]: I put the generated files back into git for CI. I took the COPY config line out of Dockerfile (equivalent of old manifests line), we'll see how that does with CI.

File: config/manager/kustomization.yaml
Comment (alanconway) [reply config/manager/kustomization.yaml]: This is tricky, I'll think about it. I really hate how kustomize and operator-sdk fail to distinguish between source and generated files. We could reset images names in a "pre-commit" target but people can forget to use it. We could have a not-checked-in set of "shadow" directories for dev work, but then kustomize changes that should be checked in may get lost. Hmm.

File: config/manifests/bases/cluster-logging.clusterserviceversion.yaml
Comment (jcantrill) [inline config/manifests/bases/cluster-logging.clusterserviceversion.yaml:96]: This may break upgrade without explicitly using "replace". My work with MOLO to modernize the deployment led me to:
* package name must be consistent for one operator to supersede another; the package name can not vary by operator name can
* the one operator (new) must explicitly replace the other (old) via replaces in the CSV)

The last point would allow us to change from "clusterlogging" to "cluster-logging".  Replaces is good practice (I believe) but we don't use it to date

File: config/manifests/bases/cluster-logging.clusterserviceversion.yaml
Comment (jcantrill) [inline config/manifests/bases/cluster-logging.clusterserviceversion.yaml:221]: Is this not an indicator of the kube APIs we support?  Don't we need this to ensure we don't get deployed on a non-supporting cluster?

File: config/overlays/release/deployment_patch.yaml
Comment (jcantrill) [inline config/overlays/release/deployment_patch.yaml:17]: 1.1

File: config/rbac/role.yaml
Comment (jcantrill) [inline config/rbac/role.yaml:7]: I think this is better but does it break upgrade?

File: config/rbac/service_account.yaml
Comment (jcantrill) [inline config/rbac/service_account.yaml:4]: upgrade issue?

File: test/functional/collection/suite_test.go
Comment (jcantrill) [inline test/functional/collection/suite_test.go:13]: This is configurable as a switch in the make file

File: test/framework/functional/output_elasticsearch7.go
Comment (jcantrill) [inline test/framework/functional/output_elasticsearch7.go:27]: any reason this shouldn't be always?

File: manifests/DEPRECATED.md
Comment (jcantrill) [inline manifests/DEPRECATED.md:1]: I believe our only restriction is us... on the testing side.  we currently build the bundle in CI

File: hack/svt/README.md
Comment (jcantrill) [inline hack/svt/README.md:28]: Do we have a way to mimic this... This is useful when testing locally

File: hack/fluent-receiver-template.yaml
Comment (jcantrill) [inline hack/fluent-receiver-template.yaml:104]: Why change from always?

File: config/manifests/bases/cluster-logging.clusterserviceversion.yaml
Comment (alanconway) [reply config/manifests/bases/cluster-logging.clusterserviceversion.yaml:96]: I'll go back to the way it was. We may want to make the hyphenation more consistent in future, but its best not to mix it in with this commit.

File: config/manifests/bases/cluster-logging.clusterserviceversion.yaml
Comment (alanconway) [reply config/manifests/bases/cluster-logging.clusterserviceversion.yaml:221]: I'll restore it for now. I think we should be checking API versions not minKubeVersions, I'll open a separate issue.

File: config/overlays/release/deployment_patch.yaml
Comment (alanconway) [reply config/overlays/release/deployment_patch.yaml:17]: +1

File: config/rbac/role.yaml
Comment (alanconway) [reply config/rbac/role.yaml:7]: I'm reverting, all the clusterlogging => cluster-logging changes. It is more consistent but best not mix it into this commit.

File: config/rbac/service_account.yaml
Comment (alanconway) [reply config/rbac/service_account.yaml:4]: Will revert.

File: test/functional/collection/suite_test.go
Comment (alanconway) [reply test/functional/collection/suite_test.go:13]: Yes. Reason I did this is some tests don't understand ginkgo flags so if you 'go test' a group of tests that includes non-ginkgo tests with the ginkgo flag, it fails. On thing I'm aiming for is to be able to say "go test ..." with whatever set of tests you want. If this bothers you I can look for a different way to do it.

File: test/framework/functional/output_elasticsearch7.go
Comment (alanconway) [reply test/framework/functional/output_elasticsearch7.go:27]: Slows the build, but more importantly if you run the tests intensively as I have been doing, you hit docker's rate limiting and get flakes. Everywhere else I've switched to quay.io images, is there a quay equivalent of this one?

File: manifests/DEPRECATED.md
Comment (alanconway) [reply manifests/DEPRECATED.md:1]: Right, I kept this to avoid breaking CI. Once this is merged we can switch CI also to the operator SDK method.

File: hack/fluent-receiver-template.yaml
Comment (alanconway) [reply hack/fluent-receiver-template.yaml:104]: Slows the build. For images that we don't build, that are properly tagged and versioned in quay or docker, it seems un-necesssary to pull every time. For images we do build I'm pulling always. If you object I can revert and we can reconsider later.

File: test/functional/collection/suite_test.go
Comment (jcantrill) [reply test/functional/collection/suite_test.go:13]: I think generally "go test" with ginkgo is always a local dev problem.  To me this means, you dev, just need to understand the implications of "slow test" and we strive to solve by adding a higher level flag in the makefile.  That being said, I'm not hard pressed for you to remove this; it was just a suggestion to resolve not needing to touch so many files.

File: test/framework/functional/output_elasticsearch7.go
Comment (jcantrill) [reply test/framework/functional/output_elasticsearch7.go:27]: My ask here is regarding the pull policy which is a kubernetes thing.  I prefer "always" for the dev experience so you make sure your changes are actually being tested.  Maybe since this is an image upon which we rely and don't devel we are OK here

File: README.md
Comment (jcantrill) [inline README.md:34]: Shouldn't this still be plural?

File: README.md
Comment (jcantrill) [inline README.md:38]: This API no longer exists in the code base

File: hack/fluent-receiver-template.yaml
Comment (alanconway) [reply hack/fluent-receiver-template.yaml:104]: The container engine will only pull the image if the digest changes, but it has to get the digest every time the container runs, so it still has some impact. Maybe it's not worth bothering about, I should do some measurements. 
https://kubernetes.io/docs/concepts/containers/images/
```
Always
    every time the kubelet launches a container, the kubelet queries the container image registry to resolve the name to an image [digest](https://docs.docker.com/engine/reference/commandline/pull/#pull-an-image-by-digest-immutable-identifier). If the kubelet has a container image with that exact digest cached locally, the kubelet uses its cached image; otherwise, the kubelet pulls the image with the resolved digest, and uses that image to launch the container.
```

**Comment by alanconway** [comment]
Closing this PR as it is being introduced in pieces by other PRs e.g. https://github.com/openshift/cluster-logging-operator/pull/1573

**Pull request closed without merging**
